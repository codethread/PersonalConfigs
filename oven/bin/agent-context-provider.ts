// :module: Provides contextual AGENTS.md documentation to Claude Code based on file reads with session-based deduplication

import {existsSync, readFileSync, unlinkSync, writeFileSync} from "fs";
import {dirname, join, resolve} from "path";
import {parseArgs} from "util";
import type {HookInput} from "../shared/claude-hooks";
import {report, reportError} from "../shared/report";

function showHelp() {
	console.log(`
agent-context-provider - Provide contextual AGENTS.md documentation to Claude Code

Usage: agent-context-provider <command> [options]

Commands:
  session-start    Initialize session tracking (called from SessionStart hook)
  session-end      Clean up session tracking (called from SessionEnd hook)
  read            Process file read and provide context (called from PostToolUse Read hook)

Options:
  --help, -h      Show this help message

This tool is designed to be called from Claude Code hooks to automatically provide
relevant AGENTS.md documentation when files are read, with intelligent deduplication
to avoid spamming Claude with repeated documentation.
`);
	process.exit(0);
}

export interface SessionState {
	sessionId: string;
	seenAgentsFiles: Set<string>;
	projectRoot: string;
}

export interface AgentContextOptions {
	command: "session-start" | "session-end" | "read";
	sessionId?: string;
	filePath?: string;
	projectRoot?: string;
}

async function main() {
	const {positionals, values} = parseArgs({
		args: Bun.argv.slice(2),
		options: {
			help: {type: "boolean", short: "h"},
		},
		allowPositionals: true,
	});

	if (values.help) {
		showHelp();
	}

	if (positionals.length === 0) {
		console.error("Error: command is required");
		showHelp();
	}

	const command = positionals[0] as "session-start" | "session-end" | "read";

	if (!["session-start", "session-end", "read"].includes(command)) {
		console.error("Error: Invalid command. Must be session-start, session-end, or read");
		process.exit(1);
	}

	try {
		// Read stdin data like claude-code-logger does
		const stdinData = await Bun.stdin.text();
		let hookInput: Partial<HookInput> = {};

		if (stdinData.trim()) {
			hookInput = JSON.parse(stdinData) as HookInput;
		}

		// Extract relevant data based on hook type
		const sessionId = hookInput.session_id;
		let filePath: string | undefined;
		const projectRoot = hookInput.cwd;

		// For read command, extract file path from PostToolUse Read events
		if (command === "read" && hookInput.hook_event_name === "PostToolUse") {
			const toolInput = (hookInput as any).tool_input;
			filePath = toolInput?.file_path;
		}

		const result = await agentContextProviderLib({
			command,
			sessionId,
			filePath,
			projectRoot,
		});

		report(result);
	} catch (err) {
		reportError(err);
		process.exit(1);
	}
}

export async function agentContextProviderLib(options: AgentContextOptions) {
	switch (options.command) {
		case "session-start":
			return await handleSessionStart(options);
		case "session-end":
			return await handleSessionEnd(options);
		case "read":
			return await handleRead(options);
		default:
			throw new Error(`Unknown command: ${options.command}`);
	}
}

async function handleSessionStart(
	options: AgentContextOptions,
): Promise<{success: boolean; message: string}> {
	if (!options.sessionId) {
		throw new Error("Session ID is required for session-start");
	}

	if (!options.projectRoot) {
		throw new Error("Project root is required for session-start");
	}

	const sessionState: SessionState = {
		sessionId: options.sessionId,
		seenAgentsFiles: new Set(),
		projectRoot: options.projectRoot,
	};

	const sessionFile = getSessionFilePath(options.sessionId);
	writeSessionState(sessionFile, sessionState);

	// Provide introductory context about the automatic documentation system
	const introMessage = `<project-context>
IMPORTANT: Ensure you read related documentation when working in nested areas of the codebase`;

	// Find all AGENTS.md files in the project to list them
	const agentsFiles: string[] = [];
	function findAllAgentsFiles(dir: string) {
		const glob = new Bun.Glob("**/AGENTS.md");
		for (const file of glob.scanSync({cwd: dir, absolute: false})) {
			if (!file.includes("node_modules") && !file.includes(".git")) {
				agentsFiles.push(file);
			}
		}
	}

	try {
		findAllAgentsFiles(options.projectRoot);
	} catch (_e) {
		// Ignore errors in finding files
	}

	if (agentsFiles.length > 0) {
		const sortedFiles = agentsFiles.sort();
		console.log(`${introMessage}
Project documentation:
${sortedFiles.map((f) => `- ${f}`).join("\n")}
</project-context>`);
	}

	return {
		success: true,
		message: `Initialized agent context session: ${options.sessionId}`,
	};
}

async function handleSessionEnd(
	options: AgentContextOptions,
): Promise<{success: boolean; message: string}> {
	if (!options.sessionId) {
		throw new Error("Session ID is required for session-end");
	}

	const sessionFile = getSessionFilePath(options.sessionId);
	if (existsSync(sessionFile)) {
		unlinkSync(sessionFile);
	}

	return {
		success: true,
		message: `Cleaned up agent context session: ${options.sessionId}`,
	};
}

async function handleRead(
	options: AgentContextOptions,
): Promise<{success: boolean; agentsFound: number}> {
	if (!options.sessionId) {
		throw new Error("Session ID is required for read command");
	}

	if (!options.filePath) {
		throw new Error("File path is required for read command");
	}

	const sessionFile = getSessionFilePath(options.sessionId);
	if (!existsSync(sessionFile)) {
		// Session not initialized, skip processing
		return {success: true, agentsFound: 0};
	}

	const sessionState = readSessionState(sessionFile);
	const agentsFiles = findAgentsFiles(options.filePath, sessionState.projectRoot);

	let agentsFound = 0;
	const newAgentsContent: string[] = [];

	for (const agentsFile of agentsFiles) {
		if (!sessionState.seenAgentsFiles.has(agentsFile)) {
			const content = readFileSync(agentsFile, "utf8");
			const relativePath = agentsFile.replace(sessionState.projectRoot, "").replace(/^\//, "");

			newAgentsContent.push(`<agent-documentation-context file="${relativePath}">
${content}
</agent-documentation-context>`);

			sessionState.seenAgentsFiles.add(agentsFile);
			agentsFound++;
		}
	}

	if (newAgentsContent.length > 0) {
		// Output just the XML-tagged content - the SessionStart already explained what this is
		const contextMessage = newAgentsContent.join("\n\n");

		// Use stderr with exit code 2 to ensure Claude sees the context
		// Exit code 2 for PostToolUse shows stderr to Claude
		console.error(contextMessage);

		// Update session state
		writeSessionState(sessionFile, sessionState);

		// Exit with code 2 to make this a "blocking" event that Claude will see
		// This ensures the context is actually delivered to Claude
		process.exit(2);
	}

	return {success: true, agentsFound};
}

function findAgentsFiles(filePath: string, projectRoot: string): string[] {
	const agentsFiles: string[] = [];
	const normalizedProjectRoot = resolve(projectRoot);
	let currentDir = resolve(dirname(filePath));

	// Traverse up the directory tree until we reach the project root
	while (currentDir.startsWith(normalizedProjectRoot)) {
		const agentsPath = join(currentDir, "AGENTS.md");
		if (existsSync(agentsPath)) {
			agentsFiles.push(agentsPath);
		}

		// Stop if we've reached the project root
		if (currentDir === normalizedProjectRoot) {
			break;
		}

		const parentDir = dirname(currentDir);
		if (parentDir === currentDir) {
			// Reached filesystem root
			break;
		}
		currentDir = parentDir;
	}

	// Return in order from deepest to shallowest (closest to file first)
	return agentsFiles;
}

function getSessionFilePath(sessionId: string): string {
	return join("/tmp", `claude-agent-context-${sessionId}.json`);
}

function readSessionState(sessionFile: string): SessionState {
	const data = readFileSync(sessionFile, "utf8");
	const parsed = JSON.parse(data);

	// Convert seenAgentsFiles array back to Set
	return {
		...parsed,
		seenAgentsFiles: new Set(parsed.seenAgentsFiles || []),
	};
}

function writeSessionState(sessionFile: string, sessionState: SessionState): void {
	// Convert Set to array for JSON serialization
	const serializable = {
		...sessionState,
		seenAgentsFiles: Array.from(sessionState.seenAgentsFiles),
	};

	writeFileSync(sessionFile, JSON.stringify(serializable, null, 2));
}

// Only run if executed directly
if (import.meta.main) {
	main();
}
