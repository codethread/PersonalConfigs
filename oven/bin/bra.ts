import {$} from "bun";
import {fzf} from "./shared/fzf";

// try to remove remote/<blah> if it exists
const remoteRegex = /remotes\/[^/]*\//;

function showHelp() {
	console.log(`
bra - Git branch switcher with fzf

Usage: bra [options]

Options:
  -h, --help      Show this help message

Description:
  Interactive git branch switcher using fzf. Shows all local and remote branches,
  allows fuzzy searching, and switches to the selected branch.

Examples:
  bra             # Launch interactive branch selector
`);
}

async function main() {
	const args = process.argv.slice(2);

	// Check for help flag
	if (args.includes("-h") || args.includes("--help")) {
		showHelp();
		process.exit(0);
	}

	try {
		const branches = (await $`git branch --all | grep -v HEAD`.text("utf-8")).split("\n");

		const behind = branches.filter((b) => b.startsWith("+"));
		const out = await fzf(branches, {tmux: true});

		const branch = out.trim().replace(remoteRegex, "").replace("+ ", "").replace("* ", "");

		// TODO: this also means checkedout in a worktree
		if (false && behind.length > 1) {
			console.log("the following are behind their remote");
			console.log(behind);
			console.log("");
		}
		await $`git checkout ${branch}`;
	} catch (e) {
		console.error(e);
		process.exit(1);
	}
}

main();
