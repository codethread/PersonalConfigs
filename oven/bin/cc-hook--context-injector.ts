// :module: Claude Code hook that provides contextual AGENTS.md documentation based on file reads with session-based deduplication

import {existsSync, readFileSync, unlinkSync, writeFileSync} from "fs";
import {dirname, join, resolve} from "path";
import {parseArgs} from "util";
import type {HookInput} from "../shared/claude-hooks";
import {report, reportError} from "../shared/report";

function showHelp() {
	console.log(`
cc-hook--context-injector - Claude Code hook that provides contextual AGENTS.md documentation

Usage: cc-hook--context-injector <command> [options]

Commands:
  session-start    Initialize session tracking (called from SessionStart hook)
  session-end      Clean up session tracking (called from SessionEnd hook)
  read            Process file read and provide context (called from PostToolUse Read hook)

Options:
  --help, -h      Show this help message

This tool is designed to be called from Claude Code hooks to automatically provide
relevant AGENTS.md documentation when files are read, with intelligent deduplication
to avoid spamming Claude with repeated documentation.
`);
	process.exit(0);
}

export interface SessionState {
	sessionId: string;
	seenAgentsFiles: Set<string>;
	projectRoot: string;
}

export interface AgentContextOptions {
	command: "session-start" | "session-end" | "read";
	sessionId?: string;
	filePath?: string;
	projectRoot?: string;
}

async function main() {
	const {positionals, values} = parseArgs({
		args: Bun.argv.slice(2),
		options: {
			help: {type: "boolean", short: "h"},
		},
		allowPositionals: true,
	});

	if (values.help) {
		showHelp();
	}

	if (positionals.length === 0) {
		console.error("Error: command is required");
		showHelp();
	}

	const command = positionals[0] as "session-start" | "session-end" | "read";

	if (!["session-start", "session-end", "read"].includes(command)) {
		console.error("Error: Invalid command. Must be session-start, session-end, or read");
		process.exit(1);
	}

	try {
		// Read stdin data like claude-code-logger does
		const stdinData = await Bun.stdin.text();
		let hookInput: Partial<HookInput> = {};

		if (stdinData.trim()) {
			hookInput = JSON.parse(stdinData) as HookInput;
		}

		// Extract relevant data based on hook type
		const sessionId = hookInput.session_id;
		let filePath: string | undefined;
		const projectRoot = hookInput.cwd;

		// For read command, extract file path from PostToolUse Read events
		if (command === "read" && hookInput.hook_event_name === "PostToolUse") {
			const toolInput = (hookInput as any).tool_input;
			filePath = toolInput?.file_path;
		}

		const result = await ccHookContextInjectorLib({
			command,
			sessionId,
			filePath,
			projectRoot,
		});

		// Handle special case for session-start contextOutput
		if (
			command === "session-start" &&
			result &&
			typeof result === "object" &&
			"contextOutput" in result &&
			result.contextOutput
		) {
			// Output context directly to stdout for Claude to see
			console.log(result.contextOutput);
		}

		report(result);
	} catch (err) {
		reportError(err);
		process.exit(1);
	}
}

export async function ccHookContextInjectorLib(options: AgentContextOptions): Promise<HandlerResult> {
	switch (options.command) {
		case "session-start":
			return await handleSessionStart(options);
		case "session-end":
			return await handleSessionEnd(options);
		case "read":
			return await handleRead(options);
		default:
			throw new Error(`Unknown command: ${options.command}`);
	}
}

type SessionStartResult = {
	success: boolean;
	message: string;
	contextOutput?: string;
};

type SessionEndResult = {
	success: boolean;
	message: string;
};

type ReadResult = {
	success: boolean;
	agentsFound: number;
};

type HandlerResult = SessionStartResult | SessionEndResult | ReadResult;

async function handleSessionStart(options: AgentContextOptions): Promise<SessionStartResult> {
	if (!options.sessionId) {
		throw new Error("Session ID is required for session-start");
	}

	if (!options.projectRoot) {
		throw new Error("Project root is required for session-start");
	}

	const sessionState: SessionState = {
		sessionId: options.sessionId,
		seenAgentsFiles: new Set(),
		projectRoot: options.projectRoot,
	};

	const sessionFile = getSessionFilePath(options.sessionId);
	writeSessionState(sessionFile, sessionState);

	// Provide introductory context about the automatic documentation system
	const introMessage = `<project-context>
IMPORTANT: Ensure you read related documentation when working in nested areas of the codebase`;

	// Find all README.md files in the project to list them (AGENTS.md files are injected contextually)
	const readmeFiles: string[] = [];
	function findAllReadmeFiles(dir: string) {
		const readmeGlob = new Bun.Glob("**/README.md");
		for (const file of readmeGlob.scanSync({cwd: dir, absolute: false})) {
			if (!file.includes("node_modules") && !file.includes(".git")) {
				readmeFiles.push(file);
			}
		}
	}

	try {
		findAllReadmeFiles(options.projectRoot);
	} catch (_e) {
		// Ignore errors in finding files
	}

	let contextOutput: string | undefined;
	if (readmeFiles.length > 0) {
		const sortedReadmeFiles = readmeFiles.sort();
		contextOutput = `${introMessage}
Project documentation:
${sortedReadmeFiles.map((f) => `- ${f}`).join("\n")}
</project-context>`;
	}

	return {
		success: true,
		message: `Initialized agent context session: ${options.sessionId}`,
		contextOutput,
	};
}

async function handleSessionEnd(options: AgentContextOptions): Promise<SessionEndResult> {
	if (!options.sessionId) {
		throw new Error("Session ID is required for session-end");
	}

	const sessionFile = getSessionFilePath(options.sessionId);
	if (existsSync(sessionFile)) {
		unlinkSync(sessionFile);
	}

	return {
		success: true,
		message: `Cleaned up agent context session: ${options.sessionId}`,
	};
}

async function handleRead(options: AgentContextOptions): Promise<ReadResult> {
	if (!options.sessionId) {
		throw new Error("Session ID is required for read command");
	}

	if (!options.filePath) {
		throw new Error("File path is required for read command");
	}

	const sessionFile = getSessionFilePath(options.sessionId);
	if (!existsSync(sessionFile)) {
		// Session not initialized, skip processing
		return {success: true, agentsFound: 0};
	}

	const sessionState = readSessionState(sessionFile);
	const agentsFiles = findAgentsFiles(options.filePath, sessionState.projectRoot);

	let agentsFound = 0;
	const newAgentsContent: string[] = [];

	for (const agentsFile of agentsFiles) {
		if (!sessionState.seenAgentsFiles.has(agentsFile)) {
			const content = readFileSync(agentsFile, "utf8");
			const relativePath = agentsFile.replace(sessionState.projectRoot, "").replace(/^\//, "");

			newAgentsContent.push(`<agent-documentation-context file="${relativePath}">
${content}
</agent-documentation-context>`);

			sessionState.seenAgentsFiles.add(agentsFile);
			agentsFound++;
		}
	}

	if (newAgentsContent.length > 0) {
		// Output just the XML-tagged content - the SessionStart already explained what this is
		const contextMessage = newAgentsContent.join("\n\n");

		// Use stderr with exit code 2 to ensure Claude sees the context
		// Exit code 2 for PostToolUse shows stderr to Claude
		console.error(contextMessage);

		// Update session state
		writeSessionState(sessionFile, sessionState);

		// Exit with code 2 to make this a "blocking" event that Claude will see
		// This ensures the context is actually delivered to Claude
		process.exit(2);
	}

	return {success: true, agentsFound};
}

function findAgentsFiles(filePath: string, projectRoot: string): string[] {
	const agentsFiles: string[] = [];
	const normalizedProjectRoot = resolve(projectRoot);
	let currentDir = resolve(dirname(filePath));

	// Traverse up the directory tree until we reach the project root
	while (currentDir.startsWith(normalizedProjectRoot)) {
		const agentsPath = join(currentDir, "AGENTS.md");
		if (existsSync(agentsPath)) {
			agentsFiles.push(agentsPath);
		}

		// Stop if we've reached the project root
		if (currentDir === normalizedProjectRoot) {
			break;
		}

		const parentDir = dirname(currentDir);
		if (parentDir === currentDir) {
			// Reached filesystem root
			break;
		}
		currentDir = parentDir;
	}

	// Return in order from deepest to shallowest (closest to file first)
	return agentsFiles;
}

function getSessionFilePath(sessionId: string): string {
	return join("/tmp", `claude-agent-context-${sessionId}.json`);
}

function readSessionState(sessionFile: string): SessionState {
	const data = readFileSync(sessionFile, "utf8");
	const parsed = JSON.parse(data);

	// Convert seenAgentsFiles array back to Set
	return {
		...parsed,
		seenAgentsFiles: new Set(parsed.seenAgentsFiles || []),
	};
}

function writeSessionState(sessionFile: string, sessionState: SessionState): void {
	// Convert Set to array for JSON serialization
	const serializable = {
		...sessionState,
		seenAgentsFiles: Array.from(sessionState.seenAgentsFiles),
	};

	writeFileSync(sessionFile, JSON.stringify(serializable, null, 2));
}

// Only run if executed directly
if (import.meta.main) {
	main();
}
