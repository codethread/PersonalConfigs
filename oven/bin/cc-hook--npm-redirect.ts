// :module: Claude Code hook that redirects npm/npx/node commands to the detected package manager (pnpm, bun, yarn)

import {existsSync} from "fs";
import {dirname, isAbsolute, resolve} from "path";
import {parseArgs} from "util";
import type {HookInput, PreToolUseInput} from "../shared/claude-hooks";
import {reportError} from "../shared/report";

function showHelp() {
	console.log(`
cc-hook--npm-redirect - Claude Code hook that redirects npm/npx/node commands to detected package manager

Usage: cc-hook--npm-redirect [options]

Options:
  --help, -h      Show this help message

This hook is designed to be called from Claude Code PreToolUse Bash hooks to automatically
redirect npm/npx/node commands to the appropriate package manager (pnpm, bun, yarn) based on
lock files in the project directory.

The hook will redirect:
- npm/npx commands to the detected package manager (pnpm, bun, yarn)
- node commands to bun in Bun projects (with compatibility warnings)

The hook reads stdin for the Claude Code hook input and will block the command with exit code 2
if a package manager mismatch is detected, suggesting the correct command to use instead.
`);
	process.exit(0);
}

export interface PackageManagerRedirectOptions {
	command: string;
	cwd?: string;
}

export interface NpmRedirectResult {
	shouldBlock: boolean;
	blockedCommand: string | null;
	suggestedCommand: string | null;
	detectedPackageManager: string;
	warning?: string;
}

async function main() {
	const {values} = parseArgs({
		args: Bun.argv.slice(2),
		options: {
			help: {type: "boolean", short: "h"},
		},
	});

	if (values.help) {
		showHelp();
	}

	try {
		// Read stdin data from Claude Code hook
		const stdinData = await Bun.stdin.text();
		if (!stdinData.trim()) {
			process.exit(0);
		}

		const hookInput = JSON.parse(stdinData) as HookInput;

		// Only process PreToolUse Bash events
		if (hookInput.hook_event_name !== "PreToolUse") {
			process.exit(0);
		}

		const preToolUseInput = hookInput as PreToolUseInput;
		if (preToolUseInput.tool_name !== "Bash") {
			process.exit(0);
		}

		const command = preToolUseInput.tool_input?.command || "";
		if (!command) {
			process.exit(0);
		}

		const result = await ccHookNpmRedirectLib({
			command,
			cwd: hookInput.cwd,
		});

		if (result.shouldBlock) {
			// Send error message to stderr for Claude to see
			let errorMessage = `Error: Use '${result.suggestedCommand}' instead of '${result.blockedCommand}'`;
			if (result.warning) {
				errorMessage += `\n${result.warning}`;
			}
			console.error(errorMessage);
			// Exit with code 2 to signal Claude to correct
			process.exit(2);
		}

		// Command is allowed, exit successfully
		process.exit(0);
	} catch (err) {
		reportError(err);
		process.exit(1);
	}
}

export async function ccHookNpmRedirectLib(
	options: PackageManagerRedirectOptions,
): Promise<NpmRedirectResult> {
	const {command, cwd} = options;

	// Detect the appropriate package manager based on lock files
	const detectedPM = detectPackageManager(command, cwd);

	// Check for package manager commands
	const result = checkPackageManagerMismatch(command, detectedPM);

	return {
		shouldBlock: result.blockedCommand !== null,
		blockedCommand: result.blockedCommand,
		suggestedCommand: result.suggestedCommand,
		detectedPackageManager: detectedPM,
		warning: result.warning,
	};
}

function isInsideQuotes(text: string, position: number): boolean {
	let inSingleQuote = false;
	let inDoubleQuote = false;
	let inBacktick = false;

	for (let i = 0; i < position; i++) {
		const char = text[i];
		const prevChar = i > 0 ? text[i - 1] : "";

		// Skip escaped quotes
		if (prevChar === "\\") {
			continue;
		}

		if (char === "'" && !inDoubleQuote && !inBacktick) {
			inSingleQuote = !inSingleQuote;
		} else if (char === '"' && !inSingleQuote && !inBacktick) {
			inDoubleQuote = !inDoubleQuote;
		} else if (char === "`" && !inSingleQuote && !inDoubleQuote) {
			inBacktick = !inBacktick;
		}
	}

	return inSingleQuote || inDoubleQuote || inBacktick;
}

function checkPackageManagerMismatch(
	command: string,
	detectedPM: string,
): {blockedCommand: string | null; suggestedCommand: string | null; warning?: string} {
	// Match package manager commands at word boundaries
	const npmPattern = /\bnpm\s+/g;
	const npxPattern = /\bnpx\s+/g;
	const yarnPattern = /\byarn\s+/g;
	const pnpmPattern = /\bpnpm\s+/g;
	const nodePattern = /\bnode\s+/g;

	let blockedCommand: string | null = null;
	let suggestedCommand: string | null = null;
	let warning: string | undefined;

	// Check npm pattern
	let match = npmPattern.exec(command);
	if (match && !isInsideQuotes(command, match.index) && detectedPM !== "npm") {
		blockedCommand = command;
		suggestedCommand = command.replace(/\bnpm\s+/g, `${detectedPM} `);
		return {blockedCommand, suggestedCommand, warning};
	}

	// Reset regex
	npmPattern.lastIndex = 0;

	// Check npx pattern
	match = npxPattern.exec(command);
	if (match && !isInsideQuotes(command, match.index) && detectedPM !== "npm") {
		blockedCommand = command;
		const execCommand = getExecutorCommand(detectedPM);
		suggestedCommand = command.replace(/\bnpx\s+/g, `${execCommand} `);
		return {blockedCommand, suggestedCommand, warning};
	}

	// Reset regex
	npxPattern.lastIndex = 0;

	// Check yarn pattern
	match = yarnPattern.exec(command);
	if (match && !isInsideQuotes(command, match.index) && detectedPM !== "yarn") {
		blockedCommand = command;
		suggestedCommand = command.replace(/\byarn\s+/g, `${detectedPM} `);
		return {blockedCommand, suggestedCommand, warning};
	}

	// Reset regex
	yarnPattern.lastIndex = 0;

	// Check pnpm pattern
	match = pnpmPattern.exec(command);
	if (match && !isInsideQuotes(command, match.index) && detectedPM !== "pnpm") {
		blockedCommand = command;
		suggestedCommand = command.replace(/\bpnpm\s+/g, `${detectedPM} `);
		return {blockedCommand, suggestedCommand, warning};
	}

	// Reset regex
	pnpmPattern.lastIndex = 0;

	// Check node pattern
	match = nodePattern.exec(command);
	if (match && !isInsideQuotes(command, match.index) && detectedPM === "bun") {
		blockedCommand = command;
		suggestedCommand = command.replace(/\bnode\s+/g, "bun ");
		warning =
			"Note: Node.js and Bun have different CLI flags. You may need to adjust any node-specific flags for Bun compatibility.";
		return {blockedCommand, suggestedCommand, warning};
	}

	return {blockedCommand, suggestedCommand, warning};
}

function getExecutorCommand(packageManager: string): string {
	switch (packageManager) {
		case "bun":
			return "bunx";
		case "pnpm":
			return "pnpm exec";
		case "yarn":
			return "yarn dlx";
		default:
			return "npx";
	}
}

function detectPackageManager(command: string, targetDir?: string): string {
	// If command has cd, extract the target directory
	let searchDir = process.cwd();
	if (targetDir) {
		searchDir = targetDir;
	} else if (command.includes("cd ")) {
		const cdMatch = command.match(/cd\s+([^\s&&]+)/);
		if (cdMatch) {
			searchDir = cdMatch[1];
			// Handle relative paths
			if (!isAbsolute(searchDir)) {
				searchDir = resolve(process.cwd(), searchDir);
			}
		}
	}

	// Walk up the directory tree to find lock files
	let currentDir = searchDir;
	while (currentDir !== dirname(currentDir)) {
		if (existsSync(resolve(currentDir, "pnpm-lock.yaml"))) {
			return "pnpm";
		}
		if (existsSync(resolve(currentDir, "bun.lock"))) {
			return "bun";
		}
		if (existsSync(resolve(currentDir, "bun.lockb"))) {
			return "bun";
		}
		if (existsSync(resolve(currentDir, "yarn.lock"))) {
			return "yarn";
		}
		if (existsSync(resolve(currentDir, "package-lock.json"))) {
			return "npm";
		}
		currentDir = dirname(currentDir);
	}
	return "npm"; // default fallback
}

// Only run if executed directly
if (import.meta.main) {
	main();
}
