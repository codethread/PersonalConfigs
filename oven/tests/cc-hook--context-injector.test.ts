import {afterEach, beforeEach, describe, expect, test} from "bun:test";
import {existsSync, mkdirSync, rmSync, unlinkSync, writeFileSync} from "fs";
import {join} from "path";
import {ccHookContextInjectorLib} from "../bin/cc-hook--context-injector";

describe("ccHookContextInjectorLib", () => {
	const testSessionId = "test-session-123";
	const testProjectRoot = "/tmp/test-project-agent-context";
	const sessionFile = join("/tmp", `claude-agent-context-${testSessionId}.json`);

	beforeEach(() => {
		// Clean up any existing test files
		if (existsSync(sessionFile)) {
			unlinkSync(sessionFile);
		}
		if (existsSync(testProjectRoot)) {
			rmSync(testProjectRoot, {recursive: true, force: true});
		}

		// Create test project structure
		mkdirSync(testProjectRoot, {recursive: true});
		mkdirSync(join(testProjectRoot, "src"), {recursive: true});
		mkdirSync(join(testProjectRoot, "src", "components"), {recursive: true});
		mkdirSync(join(testProjectRoot, "docs"), {recursive: true});

		// Create AGENTS.md files at different levels
		writeFileSync(join(testProjectRoot, "AGENTS.md"), "# Root Agent Documentation\nRoot level guidance.");
		writeFileSync(
			join(testProjectRoot, "src", "AGENTS.md"),
			"# Src Agent Documentation\nSource code guidance.",
		);
		writeFileSync(
			join(testProjectRoot, "src", "components", "AGENTS.md"),
			"# Components Agent Documentation\nComponent specific guidance.",
		);

		// Create some test files including additional README.md files
		writeFileSync(join(testProjectRoot, "README.md"), "# Test Project");
		writeFileSync(join(testProjectRoot, "src", "README.md"), "# Src Documentation");
		writeFileSync(join(testProjectRoot, "docs", "README.md"), "# Docs Section");
		writeFileSync(join(testProjectRoot, "src", "index.ts"), "export * from './components';");
		writeFileSync(join(testProjectRoot, "src", "components", "Button.tsx"), "export function Button() {}");
	});

	afterEach(() => {
		// Clean up test files
		if (existsSync(sessionFile)) {
			unlinkSync(sessionFile);
		}
		if (existsSync(testProjectRoot)) {
			rmSync(testProjectRoot, {recursive: true, force: true});
		}
	});

	describe("session-start", () => {
		test("should create session state file", async () => {
			const result = await ccHookContextInjectorLib({
				command: "session-start",
				sessionId: testSessionId,
				projectRoot: testProjectRoot,
			});

			expect(result.success).toBe(true);
			if ("message" in result) {
				expect(result.message).toContain(testSessionId);
			}

			if ("contextOutput" in result) {
				expect(result.contextOutput).toContain("Project documentation:");
			}
			expect(existsSync(sessionFile)).toBe(true);

			const sessionData = await Bun.file(sessionFile).json();
			expect(sessionData.sessionId).toBe(testSessionId);
			expect(sessionData.projectRoot).toBe(testProjectRoot);
		});

		test("should require session ID", async () => {
			await expect(
				ccHookContextInjectorLib({
					command: "session-start",
					projectRoot: testProjectRoot,
				}),
			).rejects.toThrow("Session ID is required");
		});

		test("should require project root", async () => {
			await expect(
				ccHookContextInjectorLib({
					command: "session-start",
					sessionId: testSessionId,
				}),
			).rejects.toThrow("Project root is required");
		});

		test("should include only README.md files in context output", async () => {
			const result = await ccHookContextInjectorLib({
				command: "session-start",
				sessionId: testSessionId,
				projectRoot: testProjectRoot,
			});

			expect(result.success).toBe(true);
			expect(result).toHaveProperty("contextOutput");

			const contextOutput = (result as any).contextOutput as string;
			expect(contextOutput).toContain("Project documentation:");
			expect(contextOutput).toContain("README.md");
			expect(contextOutput).toContain("src/README.md");
			expect(contextOutput).toContain("docs/README.md");
			// Should NOT contain AGENTS.md files (they are loaded automatically by Claude Code)
			expect(contextOutput).not.toContain("AGENTS.md");
			expect(contextOutput).not.toContain("src/AGENTS.md");
			expect(contextOutput).not.toContain("src/components/AGENTS.md");
		});
	});

	describe("session-end", () => {
		test("should clean up session state file", async () => {
			// First create a session
			await ccHookContextInjectorLib({
				command: "session-start",
				sessionId: testSessionId,
				projectRoot: testProjectRoot,
			});

			expect(existsSync(sessionFile)).toBe(true);

			// Then end it
			const result = await ccHookContextInjectorLib({
				command: "session-end",
				sessionId: testSessionId,
			});

			expect(result.success).toBe(true);
			if ("message" in result) {
				expect(result.message).toContain(testSessionId);
			}
			expect(existsSync(sessionFile)).toBe(false);
		});

		test("should handle missing session file gracefully", async () => {
			const result = await ccHookContextInjectorLib({
				command: "session-end",
				sessionId: testSessionId,
			});

			expect(result.success).toBe(true);
			if ("message" in result) {
				expect(result.message).toContain(testSessionId);
			}
		});

		test("should require session ID", async () => {
			await expect(
				ccHookContextInjectorLib({
					command: "session-end",
				}),
			).rejects.toThrow("Session ID is required");
		});
	});

	describe("invalid command", () => {
		test("should throw error for unknown command", async () => {
			await expect(
				ccHookContextInjectorLib({
					command: "invalid" as any,
					sessionId: testSessionId,
				}),
			).rejects.toThrow("Unknown command: invalid");
		});
	});
});
