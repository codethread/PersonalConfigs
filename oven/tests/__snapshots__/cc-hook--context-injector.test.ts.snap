// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`ccHookContextInjectorLib read should find and return AGENTS.md files for deeply nested file 1`] = `
"<agent-documentation-context file="src/components/AGENTS.md">
# Components Agent Documentation
Component specific guidance.
</agent-documentation-context>

<agent-documentation-context file="src/AGENTS.md">
# Src Agent Documentation
Source code guidance.
</agent-documentation-context>

<agent-documentation-context file="AGENTS.md">
# Root Agent Documentation
Root level guidance.
</agent-documentation-context>
"
`;

exports[`ccHookContextInjectorLib read should find AGENTS.md files for file in src directory 1`] = `
"<agent-documentation-context file="src/AGENTS.md">
# Src Agent Documentation
Source code guidance.
</agent-documentation-context>

<agent-documentation-context file="AGENTS.md">
# Root Agent Documentation
Root level guidance.
</agent-documentation-context>"
`;

exports[`ccHookContextInjectorLib read should find only root AGENTS.md for file in project root 1`] = `
"<agent-documentation-context file="AGENTS.md">
# Root Agent Documentation
Root level guidance.
</agent-documentation-context>"
`;

exports[`ccHookContextInjectorLib read should not return same AGENTS.md files twice 1`] = `
"<agent-documentation-context file="src/components/AGENTS.md">
# Components Agent Documentation
Component specific guidance.
</agent-documentation-context>

<agent-documentation-context file="src/AGENTS.md">
# Src Agent Documentation
Source code guidance.
</agent-documentation-context>

<agent-documentation-context file="AGENTS.md">
# Root Agent Documentation
Root level guidance.
</agent-documentation-context>"
`;
